### ISSUE TITLE
Question on intuition of "attention sink" and "alibi PE"

### ISSUE THREAD

**Issue by bowencohere** [mit-han-lab/streaming-llm#42]
Hi,

Thanks for the amazing work on streaming-llm. While reading the paper, I came up with this question on why applying "attention sink" also works with models with alibi position embedding.
One observation from alibi based models is that the added alibi relative bias results in very small attention scores for the tokens at the beginning of a sequence given a long sequence since the attention score follows an ~exponential decay with sequence length. Is there any explanation you can provide behind this observation as to why attention sink works under this scenario? Maybe I've missed some things from the paper, thanks very much.

**Comment by Guangxuan-Xiao** [comment]
Hi,

Thank you for your kind words and thoughtful question on StreamingLLM.

Let's address the question you have. Even though models trained with ALiBi position embeddings have an exponential decay in attention scores for the initial tokens in long sequences, they still manifest the "attention sink" phenomenon. This means that the model still dumps attention scores to these initial tokens despite the influence of the ALiBi bias diminishing their importance.

This observation underscores an insight: the attention sink phenomenon isn't solely because of a specific positional encoding method employed. Instead, it's rooted in the intrinsic nature of autoregressive language modeling and the utilization of SoftMax in attention mechanisms. 

Interestingly, our group published a paper three years ago titled "Spatten" ([link](https://arxiv.org/abs/2012.09852)). In this paper, we observed a similar attention sink phenomenon in GPT-2 model, which utilizes learned absolute positional embeddings. But we didn't further explore it at that time :).

[![Spattn Observations](https://github.com/mit-han-lab/streaming-llm/assets/40906949/d7896f96-4176-4053-83b6-3d984b0f8ba7)](https://github.com/mit-han-lab/streaming-llm/assets/40906949/d7896f96-4176-4053-83b6-3d984b0f8ba7)

I hope this clarifies the mechanism behind the attention sink in models with ALiBi embeddings.

Guangxuan

**Comment by BitCalSaul** [comment]
This phenomenon of scores accumulating in the first few tokens is quite intriguing.

I attempted to draw a connection to the paper "Vision Transformers Need Registers" (ViT).

In the case of ViT, certain unimportant tokens are automatically designated as a repository for global information, leading to their high norms. Removing these tokens would result in the loss of global information. However, if you were to remove the initial tokens at the beginning of the figure, I believe it wouldn't have a significant impact. This presents an interesting comparison with LLM.

In the case of LLM in streamingLLM, the initial tokens are automatically assigned as the primary features for future tokens. This phenomenon becomes more evident, especially in the deeper attention blocks. My hypothesis is that deeper blocks capture more global information, making it challenging to capture local information for the subsequent token predictions. Consequently, LLM automatically designates certain tokens as the collection of main features. The reason the first few tokens are chosen may be based on autoregression.

It appears that both of these papers share a common approach, where specific tokens are automatically assigned roles with distinct functions – for ViT, a collection for global information, and for LLM, a main-feature collection.
